### ISSUE TITLE
更改输入tensor维度出错

### ISSUE THREAD

**Issue by GrayRui** [alibaba/MNN#137]
在安卓demo中使用了自己转换pb模型得到的全卷积网络mnn模型，需要根据预测的图片大小调整模型的输入tensor的高和宽。
看demo使用了inputtensor.reshape(dim)，session.reshape()来更改输入tensor的维度（demo更改了通道数，我改了高和宽），所以我也这么做了。
使用的网络中有不同层输出的concat操作，结果报错说middle layer input和output不匹配，concat操作错误。但是如果我不更改tensor维度直接使用转换好的模型推理，就不会报错，结果也大致正确。
请问我这样操作有什么问题吗？是我理解错这些方法的用法了吗？

**Comment by jxt1234** [comment]
输入的原始 dim 值是?  是不是 w, h, c 的位置错了

**Comment by jxt1234** [comment]
对了，demo 不是改通道数，是改的 batch 数 和 宽高

**Comment by GrayRui** [comment]
哦，抱歉，我说错了，demo改的是batch数。。

原始网络中dim是：
dim[0] = 1  (batch)
dim[1] = 4  (channel)
dim[2] = 96 (高)
dim[3] = 192 (宽)

我的dim改成了：
dim[0] = 1  (batch)
dim[1] = 4  (channel)
dim[2] = h (高)
dim[3] = w (宽)
h,w根据计算得出的，测试的图取了96 80

**Comment by jxt1234** [comment]
应该是这个输入大小下计算出来的 concat 的输入尺寸不一致

**Comment by GrayRui** [comment]
我也想确定一下concat输入尺寸被计算成什么了，所以我想查看中间层的输出，但是有了新的疑问...
demo注释说使用session.getOutput("layer name")获取中间层输出，但是根据转换模型中输出的op->name来获取并不能获取到。
看这个getOutput调用了Session.cpp里面的getOutput方法，于是我在那里打印了一下所有mOutputs对应的名字，发现结果只有一个名字输出，看起来就是模型最终输出对应的名字，而中间层的都没有。
在转换模型的代码（tensorflowConverter.cpp）里面输出的所有op的name都是和原网络匹配的。
请问是不是我这样用法不对呢？

**Comment by jxt1234** [comment]
Pipeline.cpp 中有一个 MNN_DEBUG_TENSOR_SIZE 宏，打开看就可以
取不到 op 的问题：
1、如果 op 是一个中间的 op ，需要在 创建 session 中的 config.saveTensors 添加
2、打印一下 mnn 模型，看一下是否是被合并了，如 batchnormal, relu 会被合并到 convolution

**Comment by GrayRui** [comment]
原来是这样~
感谢大佬的耐心指教！

**Comment by GrayRui** [comment]
大佬不好意思哈，又发现问题了。。。
我的模型在concat之前，有一路是做2倍upsamle的操作的
但是我发现，无论输入的维度改成多少，这个upsamle的输出维度都是保持不变的
经过确认在tensorflow里面模型是没有问题的，请问这是bug吗？
以下是更改的三个input尺寸和它们对应计算出的upsample的尺寸

Input: 192 * 96
I/MNNJNI: ===> compute shape: block_7/upsample, [35]
  Inputs:
    1, 
    32, 
    6, 
    12, 
  Outputs:
    1, 
    32, 
    12, 
    24, 


Input: 96 * 48
I/MNNJNI: ===> compute shape: block_7/upsample, [35]
  Inputs:
    1, 
    32, 
    6, 
    3, 
  Outputs:
    1, 
    32, 
    12, 
    24, 

Input: 96 * 80
I/MNNJNI: ===> compute shape: block_7/upsample, [35]
  Inputs:
    1, 
    32, 
    6, 
    5, 
  Outputs:
    1, 
    32, 
    12, 
    24,

**Comment by jxt1234** [comment]
这应该是有 Bug 的，模型方便发我邮箱么？

**Comment by jxt1234** [comment]
看了下模型，你模型导出的时候那个 resize 的 shape 变成常数了，需要检查一下你的 tensorflow 脚本，看是不是把输入的 shape 写死了

**Comment by GrayRui** [comment]
怎么算写死了，在保存为pb模型的时候，input用shape为1 192 96 4的占位符表示了，这个会有影响吗？

def genPb():
    with tf.Session() as sess:
        img_in = tf.placeholder(dtype=tf.float32,shape=[1, 96, 192,4], name='input')
        logits, pred = narrowUNet16x.build_pred_V8_mini(img_in, H, 'test')
        sess.run(tf.global_variables_initializer())
        saver = tf.train.Saver()
        saver.restore(sess, save_path=model_path)
        frozen_graphdef = tf.graph_util.convert_variables_to_constants(sess, sess.graph_def, ["pred/pred"])
        with open("test111.pb", "wb") as f:
            f.write(frozen_graphdef.SerializeToString())

**Comment by jxt1234** [comment]
会有影响的，改成 [1, -1, -1,4] 看下

**Comment by GrayRui** [comment]
谢谢大佬，我知道问题出在哪儿了。。

原模型的上采样操作是这么做的：
trans_layer = tf.image.resize_bilinear(layer_3, shape, name='upsample')
这个resize_bilinear方法在tensorflow里面第二个参数是缩放后图像的shape，必须给定大于0的确定值（缩放后图像的大小）。这就导致占位符给定shape之后使这个resize的shape变成了常数。
